src/lib/shmem/src/allocator.rs

```diff
@@ -227,8 +212,9 @@ impl Serializer {
     /// SAFETY:
     /// * `block` must have been created from a `ShMemBlock<T>`
     /// * The backing memory must be live for the lifetime of the returned
-    ///   ShMemBlock. e.g. `ShMemBlock::deallocate` must not have been called
-    ///   on any block referencing that memory.
+    ///   ShMemBlock. i.e. the original allocated ShMemBlock must outlive the
+    ///   returned one. We can't guarantee this with normal lifetime analysis, since
+    ///   the original block may be in another process.
     pub unsafe fn deserialize<'a, T>(&'a self, block: &ShMemBlockSerialized) -> ShMemBlock<'a, T>
```

Comment:
How should the calling code enforce this, and in what situations can the backing memory be destroyed (for example, if the process crashes? the block is dropped in another process?)?

Reply:
* Shadow allocates all the blocks, and won't deallocate them before any managed code that might have deserialized the block has stopped executing.
* In managed code, before deserializing any shmem blocks, we use a prctl to tell the OS to kill the current process if the parent dies. So if Shadow exits abnormally while the managed code is still running, the OS should kill the managed code before it can access the freed shared memory. (There might be a race condition here, but I think in the worst case the managed could would segfault trying to access the freed memory)